blog post: multi-platform kubewarden (#56)
kubewarden/kubewarden.io: content/blog/2022/01/multiplatform-kubewarden.md

---
title: Multiplatform Kubewarden
authors:
    - Rafael Fernández López
date: 2022-02-01
---

The Kubewarden team is glad to announce that in the spirit of helping
Policy Authors and Cluster Administrators, the project is now
officially multiplatform.

The list of supported platforms as of now are:

- Policy Server, as [container image](https://github.com/kubewarden/policy-server/pkgs/container/policy-server):
  - `linux/amd64` (with the `musl` libc)
  - `linux/arm64` (with the `musl` libc)

- kwctl, as a [standalone binary](https://github.com/kubewarden/kwctl/releases):
  - `darwin` (`x86_64`)
  - `linux` (`aarch64`, with the `musl` libc)
  - `linux` (`x86_64`, with the `musl` libc)

>   - `windows` (`x86_64`, with `MSVC`)

Reviewer: As far as I know amd64 is equivalent to x86_64. Can we use the same name for both policy server and kwctl?

Author: Yeah... it's that docker names it as `amd64`, and our main artifact of the `policy-server` is a container image...

Author: Same happens with `arm64` vs `aarch64`. I favour `aarch64` in the binary because is what we use, while for the container image is `arm64`.

> We have tried to prioritize the same toolchain on platforms where we
> were able to do so.

Reviewer 2: ```suggestion
Our goal is to give you the same toolchain, independent of the platform you are running.
```

## Looking for platform convergence

> There are two main dependencies in the previous project layout that
> are somewhat special and sometimes problematic to converge to. One is
> `glibc`, and the other is `openssl`.

Reviewer 2: ```suggestion
Initially, we were using both `glibc` and `openssl` as dependencies in our project code.
However, these two libraries are somewhat special and sometimes problematic to converge into other platforms.

Here is a short resume of the lessons learned, sometimes the hard way, for both libraries.
```

### The libc

> `glibc` only applies to `linux` builds, so in principle is not what
> keeps us from multi platform convergence. However, we thought that
> producing a final static binary for our users in all platforms is
> interesting; specially for a CLI tool like `kwctl`, that our users
> will interact with directly in their environment.

Reviewer 2: ```suggestion
`glibc` is only used for the `linux` builds, so it would not keep us from building multi platform binaries. And we think that producing a "static binary" for the different platforms is interesting; specially for a CLI tool like `kwctl`, which our users can install directly in their environments without caring about dependencies.
```

Author: What I try to mean here is that using glibc vs musl is not helping with other platforms other than Linux. It just helps with having a standalone binary one can execute directly.

What it helps with is to have this "same" UX in all platforms, Linux included, of just downloading a binary and execute it right away.

Reviewer 3: how about listing more specific gains from the changes, besides reducing dependency graph complexity?

I would love a one-liner on how using musl allows us to cross-compile to several architectures, reducing compile time from hours in a virtualized environment to minutes when cross-compiling.

Author: Closing as discussed.

> In this spirit, statically compiling a program with `glibc` is tricky
> and can become problematic. The [`musl`](https://musl.libc.org/) libc
> is providing us with this safe, stable and straightforward static
> linking for our binaries targeting linux.

Reviewer 2: ```suggestion
However, compiling a static binary with `glibc` is tricky and can become problematic, fast. 
The [`musl libc`](https://musl.libc.org/) library is providing us with this safe, stable and straightforward static
linking for our binaries targeting linux.
```

Reviewer 2: as we are talking about multi-platform, maybe add a short sentence about the availability of musl for other platforms?

Reviewer 2: make sense. as the rewording is done, this can be closed.

### The SSL stack

The second part of the puzzle is `openssl`. It's also a complex piece
of software.

> As we have seen, we are interested in cross compiling to multiple
> Operating Systems such as Windows and Darwin.

Reviewer 2: ```suggestion
As stated above, we are interested in cross compiling for multiple
Operating Systems such as Windows and MacOS (Darwin).
```

Lucky us, a new TLS implementation has popped up in the Rust
ecosystem: [rustls](https://github.com/rustls/rustls).

> With `rustls` it's now easier to produce a final binary of `kwctl`
> that will work on all major platforms.

Reviewer 2: ```suggestion
Thanks to `rustls`, it's now easier to produce a final binary of `kwctl`
that will work on all major platforms. And all that, from the same source code.
```

## Closing

> While this decisions might seem very technical and of not-so-high
> value as other features, it's important for Kubewarden to provide a
> platform that can reach as many people as possible, while we keep the
> project as convergent as possible, trying to reduce the surface of
> critical, high-complex dependencies we use.

Reviewer 2: ```suggestion
These decisions might seem very technical, or bringing "not-so-high"
value as other features. But for the Kubewarden team, it's important we provide a
solution that can be suitable for as many people as possible.

And at the same time, we try to keep the project as convergent as possible, by reducing the number of
critical, high-complex dependencies we use in our code base.
```

> If you happen to be on an Apple Silicon Mac on darwin, [`kwctl` is not
> building yet](https://github.com/kubewarden/kwctl/issues/124). We are
> following this closely and will release the new architecture for
> `darwin` as soon as it's possible. In the meantime, you'll have to
> fallback to `rosetta` in this environment, so you can use the `x86_64`
> `kwctl` while you wait.

Reviewer 2: ```suggestion
> Note: [`kwctl` is not
building yet](https://github.com/kubewarden/kwctl/issues/124) on Apple Silicon (MacOS). We are
following this closely and will release the binary for this architecture as soon as it's possible. 
>
> In the meantime, you'll have to fallback to `rosetta` in this environment, so you can use the `x86_64` version of `kwctl`.
```

> Try Kubewarden in as many places as you want! We are eagerly looking
> for your feedback!

Reviewer 2: ```suggestion
Try [Kubewarden](https://docs.kubewarden.io/quick-start.html) today! We are eagerly looking for your feedback!
```
